### PR TITLE
Require C++17 for Halide.

### LINKED ISSUE #4216: When should we consider moving Halide to C++17?
Currently, Halide assumes/requires C++11. When should we consider moving our baseline to C++17? 

(FYI: Google recently moved to allow use of C++17 internally, so we'd be fine with this; I don't know how many other downstream users might be impacted by such a change. Would be nice to start a discussion.)

### PR DESCRIPTION
**Pull request by steven-johnson**
As of Halide 13, we require C++17 to build **or** use Halide. (We previously claimed to require C++11, although Make builds typically ended up using C++14 due to `llvm-config --cxxflags` specifying it.)

Fixes #4216

### CODE REVIEW

**Comment by steven-johnson** [comment]
The Linux failures appear to be a latent, pre-existing bug that compiler changes have revealed. Stay tuned...

**Comment by steven-johnson** [comment]
No rush to land this (we probably want to wait until after the `release/11.x` branch is created) but marking 'ready for review' so that people can review/comment at their leisure.

File: apps/bgu/CMakeLists.txt
Comment (abadams) [inline apps/bgu/CMakeLists.txt:11]: Is there no way to hide this idiocy from our users?

Comment (steven-johnson) [reply apps/bgu/CMakeLists.txt:11]: We reported this to the CMake folks and they insisted they couldn't change it without a 'policy' flag to enable it, because existing (broken) code might rely on this (insanely broken) behavior, so they aren't gonna change it anytime soon.

The alternative that most people seem to end up using is to combine sniffing of `__cplusplus` with sniffing of `_MSC_LANG`, which would work I guess, but is just so f***ing wrong that I didn't want to do it, just on principle. Maybe that's a more pragmatic approach than requiring the build files to insert extra things to work around bad decisions by both MSVC and CMake.

Comment (abadams) [reply apps/bgu/CMakeLists.txt:11]: Can we move the setting onto our side of the fence somehow, instead of in the user CMakeLists?

Comment (steven-johnson) [reply apps/bgu/CMakeLists.txt:11]: Well, yes and no.

We currently don't *need* to check __cplusplus at all (there's a reality check in Codegen_LLVM.cpp but it is just as a backstop).

If we only used C++17 features internal to libHalide, then we wouldn't need this.

If we eventually use a C++17 feature in headers exposed to user code (e.g. in templates), though, then we'd definitely have this problem (ie the user code's build system would have to set this flag).

I guess the better answer is to bite the bullet and accept that checks of that sort will have to check both __cplusplus and _MSC_LANG for the foreseeable future, which is just a sorry state of affairs.

Comment (steven-johnson) [reply apps/bgu/CMakeLists.txt]: (fix pushed)

**Review by abadams — APPROVED** [review]

**Comment by steven-johnson** [comment]
Failures are unrelated, but I want this one 100% clean before landing.

File: tutorial/lesson_01_basics.cpp
Comment (BachiLi) [inline tutorial/lesson_01_basics.cpp:6]: Since "using" Halide doesn't require c++17 -- do we need to change the compile flags in the tutorials? I don't have strong opinions, just asking.

Comment (steven-johnson) [reply tutorial/lesson_01_basics.cpp:6]: See the top comment above. Probably should keep some/all at C++11 (I just changed everything in a frenzy before thinking about it), but didn't go back and undo those changes yet to see what folks thought -- in particular, are we ready to declare it requiring C++17 for usage too? (Maybe not, but it would certainly be fine at Google.)

**Comment by alexreinking** [comment]
> NOTE: do we want to consider requiring C++17 to _use_ Halide as well? Is C++17 acceptance widespread enough (esp. among known heavy Halide users) to move in that direction?

For "modern" users, I think this is fine. MSVC 2019 supports C++17 by default. Clang and XCode have since forever. GCC is (oddly) the outlier here, having only settled the standard library ABI in 9.3, but the default Ubuntu LTS compiler is at least that new.

I can't speak for our corporate users, of course.

File: README.md
Comment (steven-johnson) [reply README.md:21]: See my top comment re: seeking opinions.

File: apps/HelloAndroid/build.gradle
Comment (steven-johnson) [reply apps/HelloAndroid/build.gradle:4]: I would be stunned if it did, but dealing with that is a separate issue.

File: src/Generator.cpp
Comment (steven-johnson) [reply src/Generator.cpp:760]: Ha, ok :-)

File: src/Generator.h
Comment (steven-johnson) [reply src/Generator.h:3114]: This file is included by client code, which (as mentioned above) shouldn't require more than C++11.

File: src/Type.h
Comment (steven-johnson) [reply src/Type.h:74]: This file is included by client code, which (as mentioned above) shouldn't require more than C++11.

Comment (steven-johnson) [reply src/Type.h:74]: Uh, I guess so. The real change here was to delete the ultra-stale comment ("We don't depend on C++11 yet").

File: src/Type.h
Comment (steven-johnson) [reply src/Type.h:21]: See above.

**Comment by steven-johnson** [comment]
To be a little more clear to reviewers about C++11 restrictions: basically, *anything* that goes into Halide.h still can't require C++14/17; since Halide.h tends to include a fair amount of theoretically-internal-only stuff, this will restrict our usage of C++17 within our internal header files.

File: src/Generator.h
Comment (alexreinking) [reply src/Generator.h:3114]: So this means we cannot use C++17 features in any header, then?

**Comment by alexreinking** [comment]
Something to be aware of: GCC 9.3 is the first version to have _non-experimental_ support for C++17. That means you cannot mix C++17 code produced by any version of GCC prior to 9.3 -- all objects / libraries must have been produced by the same compiler version.

For more info, see: https://stackoverflow.com/a/49119902/2137996

**Comment by abadams** [comment]
Perhaps we should restrict the apps (or at least some apps) to c++11 to keep ourselves honest w.r.t. Halide.h

**Comment by steven-johnson** [comment]
OK, so I:
- reverted all the apps, tests, and tutorials to build with C++11 (not C++17)
- reverted cosmetic changes to a few files that got caught in the crossfire (e.g. trailing whitespace)
- fixed a few other typos pointed out.

PTAL.

Note that per @alexreinking's comment, we may want/need to upgrade the GCC version on our bots to at least 9.3. We only use GCC on the linuxbots, and for the x86 ones it's easy (Ubuntu 20 has it preinstalled as default) but for the armbots it might be trickier. I'll investigate.

**Comment by steven-johnson** [comment]
> Note that per @alexreinking's comment, we may want/need to upgrade the GCC version on our bots to at least 9.3. 

See https://github.com/halide/Halide/issues/5645 (

**Comment by shoaibkamil** [comment]
I'm leery of making GCC 9.3+ a requirement for building code that uses Halide (it'd be a different matter if we could build Halide with GCC 9.3+ and client code using a different version of GCC).  Ubuntu 18.04 LTS has quite a large userbase.

More concretely, I'll ask devs here if this could be an issue.  Do we know what the minimum clang version is for clients using Halide with this change?

**Comment by abadams** [comment]
The decision in #4216 was based on gcc 7.1 being adequate. I don't think we should permit anything that requires gcc 9 yet.

**Comment by steven-johnson** [comment]
> The decision in #4216 was based on gcc 7.1 being adequate. I don't think we should permit anything that requires gcc 9 yet.

It's not clear to me just how dire it would be to build C++17 code with GCC 7.x and potentially mix with GCC 9.x (see https://stackoverflow.com/questions/46746878/is-it-safe-to-link-c17-c14-and-c11-objects/49119902#49119902 for a comment from a person who worked on C++17 support in GCC) -- if I read this correctly, the main issue would be if we used C++17 features (yes), compiled with GCC 7.x (yes), and then our library was used by someone using GCC 9.x (potential kablooey).

I supposed we could say "don't do that", and/or "build Halide yourself if you aren't using gcc 7.x".

Or we could back away from moving to C++17 for a while longer.

**Comment by steven-johnson** [comment]
So this is basically ready to go, with the possibly-major-or-not issue of whether GCC7 is a safe and prudent choice. Maybe let's discuss this in the next team meeting to hash out where we want to go?

**Comment by steven-johnson** [comment]
As an update to those following this: we want to do a little more investigation into compiler availability and use (wrt gcc7.x vs gcc9.x) before making a decision; due to work schedules, we're guessing this will delay a decision on this by another couple of weeks or so.

**Comment by alexreinking** [comment]
From our 3/5/21 meeting: we decided to hold off on this change until after Halide 12 is released. Halide 12 will be C++11 everywhere. Halide 13 will be C++17 everywhere (including tests, apps, etc.)

**Comment by alexreinking** [comment]
Should we resurrect this?

**Comment by steven-johnson** [comment]
> Should we resurrect this?

Yep!

File: README.md
Comment (alexreinking) [reply README.md:21]: This needs to be fixed now that we decided to go C++17 wholesale.

File: src/Generator.h
Comment (alexreinking) [reply src/Generator.h:3137]: We decided to go full C++17 so this can be reverted.

File: src/Generator.h
Comment (alexreinking) [inline src/Generator.h:3494]: Since we're going full C++17, this can be reverted, right?

**Review by alexreinking — CHANGES_REQUESTED** [review]

File: dependencies/llvm/CMakeLists.txt
Comment (steven-johnson) [inline dependencies/llvm/CMakeLists.txt:14]: @alexreinking I don't remember why this change was made -- is it possible it's just a merge hickey?

File: src/CodeGen_LLVM.cpp
Comment (steven-johnson) [inline src/CodeGen_LLVM.cpp:33]: Question: should we (also) add a similar check in Halide.h? I'm inclined to say yes, just so that downstream users aren't baffled if they get a compile error from trying to use older compilers.

File: dependencies/llvm/CMakeLists.txt
Comment (alexreinking) [reply dependencies/llvm/CMakeLists.txt:14]: I have three system wide Ubuntu packages for LLVM installed. Needed this to get it to find the right Clang when I tested the PR.

Notice this is the same thing we do when finding LLD, so it was probably an oversight when I changed that.

Comment (alexreinking) [reply dependencies/llvm/CMakeLists.txt:14]: i.e. it's a desirable change, just a bit of a rider.

File: src/CodeGen_LLVM.cpp
Comment (alexreinking) [reply src/CodeGen_LLVM.cpp:33]: I would support that, personally.

**Comment by steven-johnson** [comment]
All green (again). Are we ready to land this?

File: test/correctness/interpreter.cpp
Comment (abadams) [inline test/correctness/interpreter.cpp:136]: This file doesn't seem related?

**Comment by abadams** [comment]
No objections here.

File: README_cmake.md
Comment (alexreinking) [reply README_cmake.md:561]: @steven-johnson - can you fix this one little thing before merging? No need to wait on buildbots for this, obviously.

**Review by alexreinking — APPROVED** [review]

File: README_cmake.md
Comment (steven-johnson) [reply README_cmake.md]: done

File: test/correctness/interpreter.cpp
Comment (steven-johnson) [reply test/correctness/interpreter.cpp:136]: ...huh, weird, will revert

**Comment by abadams** [comment]
We have discussed this quite a lot at dev meetings, so I think it's good to just land at this point without waiting for explicit approvals from everyone. It can always be rolled back if someone has a serious complaint.

**Pull request merged**
